docs/ci_query.3
.\"
.\" Copyright 1998 by the Massachusetts Institute of Technology.
.\" SPDX-License-Identifier: MIT
.\"
.TH CI_QUERY 3 "24 July 1998"
.SH NAME
ci_query \- Initiate a single-question DNS query
.SH SYNOPSIS
.nf
#include <ci.h>

typedef void (*ci_callback_dnsrec)(void *arg, ci_status_t status,
                                     size_t timeouts,
                                     const ci_dns_record_t *dnsrec);

ci_status_t ci_query_dnsrec(ci_channel_t      *channel,
                                const char          *name,
                                ci_dns_class_t     dnsclass,
                                ci_dns_rec_type_t  type,
                                ci_callback_dnsrec callback,
                                void                *arg,
                                unsigned short      *qid);

typedef void (*ci_callback)(void *arg, int status,
                              int timeouts, unsigned char *abuf,
                              int alen);

void ci_query(ci_channel_t *channel, const char *name,
                int dnsclass, int type,
                ci_callback callback, void *arg);

.fi

.SH DESCRIPTION

The \fBci_query_dnsrec(3)\fP and \fBci_query(3)\fP functions initiate a
single-question DNS query on the name service channel identified by
.IR channel .
The parameter
.I name
gives the query name as a NUL-terminated C string of period-separated
labels optionally ending with a period; periods and backslashes within
a label must be escaped with a backslash.  The parameters
.I dnsclass
and
.I type
give the class and type of the query.

\fBci_query_dnsrec(3)\fP uses the ci \fBci_dns_class_t\fP and
\fBci_dns_rec_type_t\fP defined types.  However, \fBci_query(3)\fP uses
the values defined in \fB<arpa/nameser.h>\fP.

When the query is complete or has failed, the ci library will invoke
.IR callback .
Completion or failure of the query may happen immediately (even before the
return of the function call), or may happen during a later call to
\fBci_process(3)\fP or \fBci_destroy(3)\fP.

If this is called from a thread other than which the main program event loop is
running, care needs to be taken to ensure any file descriptor lists are updated
immediately within the eventloop.  When the associated callback is called,
it is called with a channel lock so care must be taken to ensure any processing
is minimal to prevent DNS channel stalls.

The callback argument
.I arg
is copied from the \fBci_query_dnsrec(3)\fP or \fBci_query(3)\fP argument
.IR arg .
The callback argument
.I status
indicates whether the query succeeded and, if not, how it failed.  It
may have any of the following values:
.TP 19
.B CI_SUCCESS
The query completed successfully.
.TP 19
.B CI_ENODATA
The query completed but contains no answers.
.TP 19
.B CI_EFORMERR
The query completed but the server claims that the query was
malformatted.
.TP 19
.B CI_ESERVFAIL
The query completed but the server claims to have experienced a
failure.  (This code can only occur if the
.B CI_FLAG_NOCHECKRESP
flag was specified at channel initialization time; otherwise, such
responses are ignored at the \fBci_send_dnsrec(3)\fP level.)
.TP 19
.B CI_ENOTFOUND
The query completed but the queried-for domain name was not found.
.TP 19
.B CI_ENOTIMP
The query completed but the server does not implement the operation
requested by the query.  (This code can only occur if the
.B CI_FLAG_NOCHECKRESP
flag was specified at channel initialization time; otherwise, such
responses are ignored at the \fBci_send_dnsrec(3)\fP level.)
.TP 19
.B CI_EREFUSED
The query completed but the server refused the query.  (This code can
only occur if the
.B CI_FLAG_NOCHECKRESP
flag was specified at channel initialization time; otherwise, such
responses are ignored at the \fBci_send_dnsrec(3)\fP level.)
.TP 19
.B CI_EBADNAME
The query name
.I name
could not be encoded as a domain name, either because it contained a
zero-length label or because it contained a label of more than 63
characters.
.TP 19
.B CI_ETIMEOUT
No name servers responded within the timeout period.
.TP 19
.B CI_ECONNREFUSED
No name servers could be contacted.
.TP 19
.B CI_ENOMEM
Memory was exhausted.
.TP 19
.B CI_ECANCELLED
The query was cancelled.
.TP 19
.B CI_EDESTRUCTION
The name service channel
.I channel
is being destroyed; the query will not be completed.
.TP 19
.B CI_ENOSERVER
The query will not be completed because no DNS servers were configured on the
channel.
.PP

The callback argument
.I timeouts
reports how many times a query timed out during the execution of the
given request.

If the query completed (even if there was something wrong with it, as
indicated by some of the above error codes), the callback argument
.I dnsrec
or
.I abuf
will be non-NULL, otherwise they will be NULL.

.SH AVAILABILITY
\fBci_query_dnsrec(3)\fP was introduced in c-ci 1.28.0.

.SH SEE ALSO
.BR ci_process (3),
.BR ci_dns_record (3)


RELEASE-NOTES.md
## c-ci version 1.34.2 - October 15 2024

This release contains a fix for downstream packages detecting the c-ci
version based on the contents of the header file rather than the
distributed pkgconf or cmake files.

## c-ci version 1.34.1 - October 9 2024

This release fixes a packaging issue.


## c-ci version 1.34.0 - October 9 2024

This is a feature and bugfix release.

Features:
* adig: read arguments from adigrc.
  [PR #856](https://github.com/c-ci/c-ci/pull/856)
* Add new pending write callback optimization via `ci_set_pending_write_cb`.
  [PR #857](https://github.com/c-ci/c-ci/pull/857)
* New function `ci_process_fds()`.
  [PR #875](https://github.com/c-ci/c-ci/pull/875)
* Failed servers should be probed rather than redirecting queries which could
  cause unexpected latency.
  [PR #877](https://github.com/c-ci/c-ci/pull/877)
* adig: rework command line arguments to mimic dig from bind.
  [PR #890](https://github.com/c-ci/c-ci/pull/890)
* Add new method for overriding network functions
  `ci_set_socket_function_ex()` to properly support all new functionality.
  [PR #894](https://github.com/c-ci/c-ci/pull/894)
* Fix regression with custom socket callbacks due to DNS cookie support.
  [PR #895](https://github.com/c-ci/c-ci/pull/895)
* ci_socket: set IP_BIND_ADDRESS_NO_PORT on ci_set_local_ip* tcp sockets
  [PR #887](https://github.com/c-ci/c-ci/pull/887)
* URI parser/writer for ci_set_servers_csv()/ci_get_servers_csv().
  [PR #882](https://github.com/c-ci/c-ci/pull/882)

Changes:
* Connection handling modularization.
  [PR #857](https://github.com/c-ci/c-ci/pull/857),
  [PR #876](https://github.com/c-ci/c-ci/pull/876)
* Expose library/utility functions to tools.
  [PR #860](https://github.com/c-ci/c-ci/pull/860)
* Remove `ci__` prefix, just use `ci_` for internal functions.
  [PR #872](https://github.com/c-ci/c-ci/pull/872)


Bugfixes:
* fix: potential WIN32_LEAN_AND_MEAN redefinition.
  [PR #869](https://github.com/c-ci/c-ci/pull/869)
* Fix googletest v1.15 compatibility.
  [PR #874](https://github.com/c-ci/c-ci/pull/874)
* Fix pkgconfig thread dependencies.
  [PR #884](https://github.com/c-ci/c-ci/pull/884)


Thanks go to these friendly people for their efforts and contributions for this
release:

* Brad House (@bradh352)
* Cristian Rodríguez (@crrodriguez)
* Georg (@tacerus)
* @lifenjoiner
* Shelley Vohr (@codebytere)
* 前进，前进，进 (@leleliu008)



src/lib/ci_query.c
/* MIT License
 *
 * Copyright (c) 1998 Massachusetts Institute of Technology
 * Copyright (c) The c-ci project and its contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice (including the next
 * paragraph) shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * SPDX-License-Identifier: MIT
 */

#include "ci_private.h"

#ifdef HAVE_NETINET_IN_H
#  include <netinet/in.h>
#endif

typedef struct {
  ci_callback_dnsrec callback;
  void                *arg;
} ci_query_dnsrec_arg_t;

static void ci_query_dnsrec_cb(void *arg, ci_status_t status,
                                 size_t                   timeouts,
                                 const ci_dns_record_t *dnsrec)
{
  ci_query_dnsrec_arg_t *qquery = arg;

  if (status != CI_SUCCESS) {
    qquery->callback(qquery->arg, status, timeouts, dnsrec);
  } else {
    size_t           ancount;
    ci_dns_rcode_t rcode;
    /* Pull the response code and answer count from the packet and convert any
     * errors.
     */
    rcode   = ci_dns_record_get_rcode(dnsrec);
    ancount = ci_dns_record_rr_cnt(dnsrec, CI_SECTION_ANSWER);
    status  = ci_dns_query_reply_tostatus(rcode, ancount);
    qquery->callback(qquery->arg, status, timeouts, dnsrec);
  }
  ci_free(qquery);
}

ci_status_t ci_query_nolock(ci_channel_t *channel, const char *name,
                                ci_dns_class_t     dnsclass,
                                ci_dns_rec_type_t  type,
                                ci_callback_dnsrec callback, void *arg,
                                unsigned short *qid)
{
  ci_status_t            status;
  ci_dns_record_t       *dnsrec = NULL;
  ci_dns_flags_t         flags  = 0;
  ci_query_dnsrec_arg_t *qquery = NULL;

  if (channel == NULL || name == NULL || callback == NULL) {
    /* LCOV_EXCL_START: DefensiveCoding */
    status = CI_EFORMERR;
    if (callback != NULL) {
      callback(arg, status, 0, NULL);
    }
    return status;
    /* LCOV_EXCL_STOP */
  }

  if (!(channel->flags & CI_FLAG_NORECURSE)) {
    flags |= CI_FLAG_RD;
  }

  status = ci_dns_record_create_query(
    &dnsrec, name, dnsclass, type, 0, flags,
    (size_t)(channel->flags & CI_FLAG_EDNS) ? channel->ednspsz : 0);
  if (status != CI_SUCCESS) {
    callback(arg, status, 0, NULL); /* LCOV_EXCL_LINE: OutOfMemory */
    return status;                  /* LCOV_EXCL_LINE: OutOfMemory */
  }

  qquery = ci_malloc(sizeof(*qquery));
  if (qquery == NULL) {
    /* LCOV_EXCL_START: OutOfMemory */
    status = CI_ENOMEM;
    callback(arg, status, 0, NULL);
    ci_dns_record_destroy(dnsrec);
    return status;
    /* LCOV_EXCL_STOP */
  }

  qquery->callback = callback;
  qquery->arg      = arg;

  /* Send it off.  qcallback will be called when we get an answer. */
  status = ci_send_nolock(channel, NULL, 0, dnsrec, ci_query_dnsrec_cb,
                            qquery, qid);

  ci_dns_record_destroy(dnsrec);
  return status;
}

ci_status_t ci_query_dnsrec(ci_channel_t *channel, const char *name,
                                ci_dns_class_t     dnsclass,
                                ci_dns_rec_type_t  type,
                                ci_callback_dnsrec callback, void *arg,
                                unsigned short *qid)
{
  ci_status_t status;

  if (channel == NULL) {
    return CI_EFORMERR;
  }

  ci_channel_lock(channel);
  status = ci_query_nolock(channel, name, dnsclass, type, callback, arg, qid);
  ci_channel_unlock(channel);
  return status;
}

void ci_query(ci_channel_t *channel, const char *name, int dnsclass,
                int type, ci_callback callback, void *arg)
{
  void *carg = NULL;

  if (channel == NULL) {
    return;
  }

  carg = ci_dnsrec_convert_arg(callback, arg);
  if (carg == NULL) {
    callback(arg, CI_ENOMEM, 0, NULL, 0); /* LCOV_EXCL_LINE: OutOfMemory */
    return;                                 /* LCOV_EXCL_LINE: OutOfMemory */
  }

  ci_query_dnsrec(channel, name, (ci_dns_class_t)dnsclass,
                    (ci_dns_rec_type_t)type, ci_dnsrec_convert_cb, carg,
                    NULL);
}
